nikola/nikola.py

```diff
@@ -1334,19 +1337,25 @@ def render_template(self, template_name, output_name, context, url_type=None):
 
         utils.makedirs(os.path.dirname(output_name))
         parser = lxml.html.HTMLParser(remove_blank_text=True)
-        doc = lxml.html.document_fromstring(data, parser)
-        self.rewrite_links(doc, src, context['lang'], url_type)
-        data = b'<!DOCTYPE html>\n' + lxml.html.tostring(doc, encoding='utf8', method='html', pretty_print=True)
+        if is_fragment:
+            doc = lxml.html.fragment_fromstring(data, parser)
+        else:
+            doc = lxml.html.document_fromstring(data, parser)
+        self.rewrite_links(doc, src, context['lang'], url_type, is_fragment=is_fragment)
+        if is_fragment:
+            data = (doc.text or '').encode('utf-8') + ''.encode('utf-8').join([lxml.html.tostring(child, encoding='utf-8', method='html') for child in doc.iterchildren()])
+        else:
+            data = b'<!DOCTYPE html>\n' + lxml.html.tostring(doc, encoding='utf8', method='html', pretty_print=True)
         with open(output_name, "wb+") as post_file:
             post_file.write(data)
 
-    def rewrite_links(self, doc, src, lang, url_type=None):
+    def rewrite_links(self, doc, src, lang, url_type=None, is_fragment=False):
         """Replace links in document to point to the right places."""
         # First let lxml replace most of them
         doc.rewrite_links(lambda dst: self.url_replacer(src, dst, lang, url_type), resolve_base_href=False)
 
         # lxml ignores srcset in img and source elements, so do that by hand
-        objs = list(doc.xpath('(*//img|*//source)'))
+        objs = list(doc.xpath('({0}//img|{0}//source)'.format('' if is_fragment else '*')))
```

Comment:
Please rewrite this as a 4-line `if/else` tree instead of abusing str.format. We should also find out what that asterisk changes (can non-fragment code work without it? Does it reach all `<img>` and `<source>` elements if the asterisk is not present?